packages/components/src/TabBar/TabBar.component.js

```diff
@@ -1,96 +1,152 @@
 /* eslint-disable react/no-find-dom-node */
-import React from 'react';
+import React, { useState, useEffect, useRef } from 'react';
 import ReactDOM from 'react-dom';
 import PropTypes from 'prop-types';
 import Tab from 'react-bootstrap/lib/Tab';
 import Nav from 'react-bootstrap/lib/Nav';
 import NavItem from 'react-bootstrap/lib/NavItem';
 import keycode from 'keycode';
+import debounce from 'lodash/debounce';
+import classnames from 'classnames';
+import theme from './TabBar.scss';
 
-class TabBar extends React.Component {
-	constructor(props) {
-		super(props);
-		this.handleSelect = this.handleSelect.bind(this);
-		this.handleKeyDown = this.handleKeyDown.bind(this);
+function TabBar(props) {
+	const tabBarContainerRef = useRef();
+	const tabBarRef = useRef();
+	const needsRefocus = useRef();
+
+	const [showDropdown, setShowDropDown] = useState(false);
+
+	function tryShowDropdown() {
+		const tabContainer = tabBarContainerRef.current;
+		if (tabContainer) {
+			// There is a TabBar, test if the right boundary point of this TabBar
+			// is closer than the right boundary point of its last nav item.
+			const lastChild = tabContainer.querySelector('li:last-child');
+			if (lastChild) {
+				if (tabContainer.getBoundingClientRect().right < lastChild.getBoundingClientRect().right) {
+					setShowDropDown(true);
+				}
+			}
+		} else {
+			// There is no TabBar, show it to test if dropdown is needed
+			setShowDropDown(false);
+		}
+	}
+
+	/**
+	 * Just show the tabBar and test if there is enough width.
+	 */
+	function showTabBarAndTest() {
+		setShowDropDown(false);
+		tryShowDropdown();
 	}
 
-	componentDidUpdate() {
-		if (!this.needsRefocus || !this.ref) {
+	useEffect(() => {
+		const resizeListener = window.addEventListener('resize', debounce(showTabBarAndTest, 200));
+		return () => window.removeEventListener('resize', resizeListener);
+	}, []);
+
+	useEffect(() => {
+		if (!needsRefocus.current || !tabBarRef.current) {
```

Comment:
Just test `needsRefocus.current` to fix the tab problem in forms (instead of `needsRefocus`).